configured MSS clamping rules under GetMSSClampTypes() table

yboaron opened the pull request:
MSS Clamping rules should be configured in different IP hooks for iptables
and nftables.

In iptables, they should be configured in mangle/postrouting hook,
while in nftables, they should be placed in filter/postrouging hook.

This PR configured MSS clamping rules under GetMSSClampTypes() hook

Fixes: https://github.com/submariner-io/submariner/issues/3355

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

yboaron committed:
Set MSS clamping rules under GetMSSClampTypes() hook

MSS clamping rules should be configured in different hooks for
iptables and nftables.
This PR configured MSS clamping rules under
GetMSSClampTypes() IP hook.

Fixes: https://github.com/submariner-io/submariner/issues/3355

Signed-off-by: Yossi Boaron <yboaron@redhat.com>

yboaron commented:
cc @sridhargaddam ^^

sridhargaddam commented on pkg/routeagent_driver/handlers/mtu/mtuhandler.go:
```diff
@@ -91,12 +91,12 @@ func (h *mtuHandler) Init(_ context.Context) error {
-		Hook:     packetfilter.ChainHookPostrouting,
+		Name:     constants.SmForwardChain,
+		Type:     packetfilter.ChainTypeFilter,
+		Hook:     packetfilter.ChainHookForward,
```
Instead of applying the MSS clamping rules in the `forward` hook, can you try in the `postrouting` hook so that it applies to both the forwarded traffic as well as traffic that is locally generated.

For example:
```
# nft add table ip mangletcpmss
# nft add chain ip mangletcpmss customchain { type filter hook postrouting priority 0 \; }
# nft add rule ip mangletcpmss customchain tcp flags syn tcp option maxseg size set rt mtu
# nft list ruleset
table ip mangletcpmss {
	chain customchain {
		type filter hook postrouting priority filter; policy accept;
		tcp flags syn tcp option maxseg size set rt mtu
	}
}
#
```

yboaron replied:
Done

yboaron commented:
Well, iptables doesn't support filter/postrouting  hook ... 

[1] 
` running [/usr/sbin/iptables -t filter -I POSTROUTING 1 -j SUBMARINER-POSTROUTING --wait 5]: exit status 1: iptables: No chain/target/match by that name.\n"`

https://upload.wikimedia.org/wikipedia/commons/3/37/Netfilter-packet-flow.svg

yboaron committed:
PacketFilter: add GetMSSClampTypes

MSS Clamping rules should be configured in different IP hooks for iptables
and nftables.

In iptables, they should be configured in mangle/postrouting hook,
while in nftables, they should be placed in filter/postrouging hook.

Signed-off-by: Yossi Boaron <yboaron@redhat.com>

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -272,6 +272,10 @@ type NamedSet interface {
 type Driver interface {
+	// MSS Clamp rules should be configured in different tables for iptables and nftables,
+	// TableTypeRoute for iptables and TableTypeFilter for nftables
+	GetMSSClampTable() TableType
```
Why don't we define a new `TableType` enum named "_MSSClamp_" instead of adding a new interface method? Then the implementation can internally decide what it should translate to.

yboaron replied:
Also thought to go with this approach first, for this approach we need two new values : 
1. new table type e.g: **TableTypeMssClamp**  , 
    and **AppendUnique , ClearChain , UpdateChainRules** APIs implementation should be updated to support this value 

2. new chain type e.g: **ChainTypeMssClamp**
 and **CreateIPHookChainIfNotExists , DeleteIPHookChain**  APIs implementation should be updated to support this value 

But in second thought, adding new GetMSSClampTable() API seems like a simpler solution to me

tpantelis commented on pkg/routeagent_driver/handlers/mtu/mtuhandler.go:
```diff
@@ -90,9 +92,16 @@ func (h *mtuHandler) Init(_ context.Context) error {
+
+	if h.tableType == packetfilter.TableTypeFilter {
+		h.chainType = packetfilter.ChainTypeFilter
+	}
```
Here the API tells which table type to use but the user has to figure out the matching chain type. How about we modify `GetMSSClampTable` to also return the chain type:

  `GetMSSClampInfo() (TableType, ChainType)`

yboaron replied:
Sounds good, I'll change it.

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -272,6 +272,10 @@ type NamedSet interface {
 type Driver interface {
+	// MSS Clamp rules should be configured in different tables for iptables and nftables,
+	// TableTypeRoute for iptables and TableTypeFilter for nftables
+	GetMSSClampTable() (TableType, ChainType)
```
```suggestion
	GetMSSClampTypes() (TableType, ChainType)
```

tpantelis committed:
Merge branch 'devel' into mss_clamp_filter_fwd
